File: roll-your-own-test-framework/python/src/roll_your_own_test_framework/test_runner.py
import inspect
from typing import Type

from .assertion_failed_exception import AssertionFailedException
from .test_result import TestResult, TestStatus
from .test_summary import TestSummary


class TestRunner:
    @staticmethod
    def run_all(test_class: Type) -> TestSummary:
        summary = TestSummary()
        instance = test_class()

        methods = inspect.getmembers(instance, predicate=inspect.ismethod)
        test_methods = [(name, method) for name, method in methods if name.startswith("test_")]

        for name, method in test_methods:
            try:
                method()
                summary.add(TestResult(name=name, status=TestStatus.PASS))
            except AssertionFailedException as e:
                summary.add(TestResult(name=name, status=TestStatus.FAIL, message=str(e)))
            except Exception as e:
                message = f"{type(e).__name__}: {e}"
                summary.add(TestResult(name=name, status=TestStatus.ERROR, message=message))

        return summary

File: roll-your-own-test-framework/csharp/src/RollYourOwnTestFramework/TestRunner.cs
using System.Reflection;

namespace RollYourOwnTestFramework;

public static class TestRunner
{
    public static TestSummary RunAll(Type testClass)
    {
        var summary = new TestSummary();
        var methods = testClass
            .GetMethods(BindingFlags.Public | BindingFlags.Instance)
            .Where(m => m.GetCustomAttribute<TestAttribute>() != null);

        foreach (var method in methods)
        {
            var instance = Activator.CreateInstance(testClass)!;
            try
            {
                method.Invoke(instance, null);
                summary.Add(new TestResult(method.Name, TestStatus.Pass));
            }
            catch (TargetInvocationException ex) when (ex.InnerException is AssertionFailedException afe)
            {
                summary.Add(new TestResult(method.Name, TestStatus.Fail, afe.Message));
            }
            catch (TargetInvocationException ex)
            {
                var inner = ex.InnerException!;
                summary.Add(new TestResult(method.Name, TestStatus.Error, $"{inner.GetType().Name}: {inner.Message}"));
            }
        }

        return summary;
    }
}
